vault/core.go

```diff
@@ -989,12 +690,20 @@ func (c *Core) Seal(token string) (retErr error) {
 		return err
 	}
 	// Attempt to use the token (decrement num_uses)
-	// If we can't, we still continue attempting the seal, so long as the token
-	// has appropriate permissions
+	// On error bail out; if the token has been revoked, bail out too
 	if te != nil {
-		if err := c.tokenStore.UseToken(te); err != nil {
+		te, err = c.tokenStore.UseToken(te)
+		if err != nil {
 			c.logger.Printf("[ERR] core: failed to use token: %v", err)
-			retErr = ErrInternalError
+			return ErrInternalError
+		}
+		if te == nil {
+			// Token is no longer valid
+			return logical.ErrPermissionDenied
+		}
+		if te.NumUses == -1 {
+			// Token needs to be revoked
+			return c.tokenStore.Revoke(te.ID)
```

Comment:
Good catch! I've changed it to defer the revocation like the normal request handling does.